packages/loopring_v3/contracts/aux/access/LoopringIOExchangeOwner.sol

```diff
@@ -139,14 +149,32 @@ contract LoopringIOExchangeOwner is SelectorBasedAccessManager, ERC1271, Drainab
             require(blockIdx < blocks.length, "INVALID_BLOCKIDX");
             ExchangeData.Block memory _block = blocks[blockIdx];
 
-            _processTxCallbacks(_block, blockCallback.txCallbacks, config.receivers);
+            _processTxCallbacks(_block, blockCallback.txCallbacks, config.receivers, preApprovedTxs[i]);
+        }
+
+        // Verify the approved transactions data against the auxiliary data in the block
+        for (uint i = 0; i < blocks.length; i++) {
+            bool[] memory _preApprovedTxs = preApprovedTxs[i];
+            ExchangeData.Block memory _block = blocks[i];
+            ExchangeData.AuxiliaryData[] memory auxiliaryData = _block.auxiliaryData;
+            for(uint j = 0; j < _block.auxiliaryData.length; j++) {
+                uint txIdx;
+                bool approved;
+                assembly {
+                    let auxOffset := mload(add(auxiliaryData, add(32, mul(32, j))))
+                    txIdx := mload(add(add(32, auxiliaryData), auxOffset))
+                    approved := mload(add(add(64, auxiliaryData), auxOffset))
+                }
+                require(_preApprovedTxs[txIdx] == approved, "PRE_APPROVED_TX_MISMATCH");
```

Comment:
Here the AMM contract verifies that non-AMM related transaction's `approved` field in auxiliary data must be false. My point is that this should really be done inside the core protocol while the data is loaded. If there is no AMM module at all, then the Loopring contract shall make sure all `approved`  fields in the auxiliary data must be false. This logic belongs to the core protocol, not the AMM contracts.

Reply:
Not sure I understand this. Do you inlclude `LoopringIOExchangeOwner` as an AMM contract?

It is a critical the data is checked, but we already depend on `LoopringIOExchangeOwner` working correctly, calling the AMM pools as expected, so already critical for security. And this is an additional critical thing.  I think it's fine doing it in `LoopringIOExchangeOwner`, there's no real need having the mechanism for pre approved transactions to be built-in. And other external contracts like the agent contract as equally critical for security. But you're certainly right that the protocol cannot be used any more without an owner contract that checks these values, which is a bit special I guess.

Reply:
Brecht, I think I'm fine with your approach. Is it possible to measure the gas reduction before we test this new version in our testing environment?

Reply:
Gas consumption in my benchmark with 2 joins went down from 622,771 gas to 498,616 gas.

Looks like the overhead for a join is now around 80,000 gas. So I think a reduction of around 40% on the join/exit part.